### PR TITLE
New custom select

### PR DESCRIPTION
**Pull request by abbiecoghlan**
### **PR Type**
Enhancement, Tests, Documentation

___

### **Description**
- Added `CustomSelectField` component with full implementation and props.
- Updated demo site to include `CustomSelectField` in configuration and stories.
- Replaced `SelectList` with `SelectField` and `CustomSelectField` in `Field` component.
- Added padding to `Field` component in demo.
- Fixed bug in `SelectField` onValueChange logic.
- Added `accessibilityState` to `TextField`.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>8 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>[component].tsx</strong><dd><code>Add padding to Field component in demo</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/demo/app/demo/[component].tsx

- Wrapped `Field` component with `Box` for padding.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/627/files#diff-22817a0745f50787d162c6e2996915d3e303b4eb671720077da432159e76418b">+11/-9</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>demoConfig.tsx</strong><dd><code>Include CustomSelectField in demo configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/demo/demoConfig.tsx

- Added `CustomSelectFieldConfiguration` to demo configuration.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/627/files#diff-dcc53f9eb9b06277a43c7bb242a3d8056ff84618e29256b7e57cca2057851744">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Banner.config.tsx</strong><dd><code>Update Banner demo configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/demo/story-config/Banner.config.tsx

- Updated `demo` property to use `BannerDemo` directly.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/627/files#diff-6f63f29001861914c2dddf6b4560bc6e7ef2f5175afe182a1c218866410b3254">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Common.ts</strong><dd><code>Update CustomSelectFieldProps interface</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/ui/src/Common.ts

- Updated `CustomSelectFieldProps` interface with new properties.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/627/files#diff-538e93b553971318d69ff6fff2ac103f295c4922e24bbe61c30897baa463bba1">+41/-6</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>CustomSelectField.tsx</strong><dd><code>Implement CustomSelectField component</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/ui/src/CustomSelectField.tsx

- Implemented `CustomSelectField` component.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/627/files#diff-048cfe177c9780e9028c59dc2d49754548e56ec764130999e237ba9c4afa10a1">+116/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Field.tsx</strong><dd><code>Replace SelectList with SelectField and CustomSelectField</code></dd></summary>
<hr>

packages/ui/src/Field.tsx

- Replaced `SelectList` with `SelectField` and `CustomSelectField`.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/627/files#diff-b179881630547b40643ddd11ccdc12624a62bf6deafcea924246db78416d25d5">+34/-38</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>TextField.tsx</strong><dd><code>Add accessibilityState to TextField</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/ui/src/TextField.tsx

- Added `accessibilityState` to `TextField`.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/627/files#diff-526ca8a54e7f6aaa340ccd6be7bb641b6ee65087e2455eef8ac1dad81625c433">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.tsx</strong><dd><code>Export CustomSelectField</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/ui/src/index.tsx

- Exported `CustomSelectField`.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/627/files#diff-b68d7ab81cb87f2517f9f0659bb424a51f0b1d7186c071447d74c07ab88c2820">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>CustomSelectField.stories.tsx</strong><dd><code>Add story for CustomSelectField component</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/demo/stories/CustomSelectField.stories.tsx

- Added story for `CustomSelectField`.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/627/files#diff-d8bd42dcbc62ccc60e5f67921a7ec9a44644d43da9496346864dfbc32de94f72">+21/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Export CustomSelectField stories</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/demo/stories/index.ts

- Exported `CustomSelectField.stories`.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/627/files#diff-590db40ce1849d210cdc32b50de4ad53b837bf58ff0e88d843cb1ff560855fdd">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>CustomSelectField.config.tsx</strong><dd><code>Add configuration for CustomSelectField component</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/demo/story-config/CustomSelectField.config.tsx

- Added configuration for `CustomSelectField`.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/627/files#diff-b5857178b95eeeb4cc987bb8b6e1e66dd3541e2b11aa1beb43e021dba14d6490">+136/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Export CustomSelectField configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/demo/story-config/index.ts

- Exported `CustomSelectField.config`.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/627/files#diff-bf78d20e924ad1719861d2a2976d9c83c8fa6ba8e5d785c8a36521b0cc8b7ea9">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Bug fix</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>SelectField.tsx</strong><dd><code>Fix bug in SelectField onValueChange logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/ui/src/SelectField.tsx

- Fixed bug in `SelectField` onValueChange logic.

</details>

  </td>
  <td><a href="https://github.com/FlourishHealth/ferns-ui/pull/627/files#diff-e18d9438d63f10fb9d7e7d0568722eaa2e25a48257b025b514645593f62bcb98">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: packages/ui/src/TextField.tsx [hunk 1]
```diff
@@ -174,6 +174,7 @@ export const TextField: FC<TextFieldProps> = ({
           }}
           accessibilityHint="Enter text here"
           accessibilityLabel="Text input field"
+          accessibilityState={{disabled}}
```
Comment (abbiecoghlan) [inline packages/ui/src/TextField.tsx:177]: I need to do a little more research on this I couldn't get my screen reader to read that the text was actually disabled when I tested it out on the demo site. It certainly doesn't hurt to have this - possibly one only works on certain screenreaders/devices. We can also update the accessibility label to share this info if needed.

File: packages/ui/src/Field.tsx [hunk 2]
```diff
@@ -186,24 +187,19 @@ export const Field = ({type, ...rest}: FieldProps) => {
...
+    if (!rest?.options) {
+      console.error("Field with type=customSelect require options");
+      return null;
+    }
+    return (
+      <CustomSelectField
+        {...(rest as CustomSelectFieldProps)}
+        onChange={(result) => {
+          rest.onChange(result);
+          rest.onBlur && rest.onBlur(result);
```
Comment (abbiecoghlan) [inline packages/ui/src/Field.tsx:199]: See thread between myself and @kjkurtz in dev about this. We will need to add onblur back in in the Field component where it was removed for other types of field, and/or move this down into the individual components. But it looks like we removed onBlur functionality in some of these during this re-working.

File: apps/demo/story-config/CustomSelectField.config.tsx [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+export const CustomSelectFieldConfiguration: DemoConfiguration = {
+  name: "Custom Select Field",
+  component: CustomSelectField,
+  related: ["SelectField", "Field", "Tap-to-edit"],
+  description:
+    "Displays a list of options using the browser’s native select and includes a custom option that renders a text field when selected and allows user to input value not included in predefined options.",
+  a11yNotes: ["The list should be labeled so that screen readers know that the list is related."],
+  category: "Component",
+  status: {
+    // TODO: ensure custom select component meets accessibility requirements
```
Comment (abbiecoghlan) [inline apps/demo/story-config/CustomSelectField.config.tsx:15]: Custom select in particular is complex in terms of accessibility. I made my best first pass at improvements, and I think we should get this merged in so it doesn't block migration and come back to it.

File: packages/ui/src/SelectList.tsx [hunk 4]
```diff
@@ -1,110 +0,0 @@
-// import React, {useContext} from "react";
```
Comment (swanka) [inline packages/ui/src/SelectList.tsx:1]: We can just remove this file if it's not needed

File: apps/demo/app/demo/[component].tsx [hunk 5]
```diff
@@ -200,15 +200,17 @@ const ComponentDemo = ({config}: {config: DemoConfiguration}) => {
...
-              label={config.demoOptions?.controls?.[prop]?.label ?? startCase(prop)}
-              {...config.demoOptions?.controls?.[prop]}
-              value={propValues[prop]}
-              onChange={(value: any) => {
-                setPropValues({...cloneDeep(propValues), [prop]: value});
-              }}
-            />
+            <Box key={prop} paddingY={2}>
+              <Field
+                // TODO: change label -> title once field props are updated
```
Comment (swanka) [inline apps/demo/app/demo/[component].tsx:205]: Let's start adding these todos to Notion here: https://www.notion.so/flourishhealth/03129716a7974bc3ba9151a187c6c717?v=d9e661b10fce4c5bbf7372b891e53220

**Review by swanka — APPROVED** [review]
Looks good to me! Definitely keep sharing what you learn about accessibility. I need to look into it because I'm not actually sure how screen readers handle nested accessibility labels on several nested dom nodes

File: packages/ui/src/SelectList.tsx (hunk 4 above)
Comment (abbiecoghlan) [reply packages/ui/src/SelectList.tsx:1]: hah yeah that's what i meant to do, my bad. done!

File: apps/demo/app/demo/[component].tsx (hunk 5 above)
Comment (abbiecoghlan) [reply apps/demo/app/demo/[component].tsx:205]: good call

**Pull request merged**